libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1641,3 +1655,145 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
 	solAssert(_expression.annotation().type, "Type of expression not set.");
 	return *_expression.annotation().type;
 }
+
+bool IRGeneratorForStatements::visit(TryStatement const& _tryStatement)
+{
+	_tryStatement.externalCall().accept(*this);
+
+	auto const trySuccessCondition = m_context.trySuccessConditionVariable(_tryStatement.externalCall());
+
+	m_code << "switch iszero(" << trySuccessCondition << ")\n";
+
+	m_code << "case 0 { // success case\n";
+	_tryStatement.clauses().front()->accept(*this);
+	m_code << "}\n";
+
+	m_code << "default { // failure case\n";
+	handleCatch(_tryStatement.clauses());
+	m_code << "}\n";
+
+	return false;
+}
+
+inline auto extractCatchClauses(vector<ASTPointer<TryCatchClause>> const& _catchClauses)
+{
+	ASTPointer<TryCatchClause> structured{};
+	ASTPointer<TryCatchClause> fallback{};
+
+	for (size_t i = 1; i < _catchClauses.size(); ++i)
+		if (_catchClauses[i]->errorName() == "Error")
+			structured = _catchClauses[i];
+		else if (_catchClauses[i]->errorName().empty())
+			fallback = _catchClauses[i];
+		else
+			solAssert(false, "");
+
+	solAssert(_catchClauses.size() == size_t(1 + (structured ? 1 : 0) + (fallback ? 1 : 0)), "");
+
+	return tuple{structured, fallback};
+}
+
+void IRGeneratorForStatements::handleCatch(vector<ASTPointer<TryCatchClause>> const& _catchClauses)
+{
+	auto const [structured, fallback] = extractCatchClauses(_catchClauses);
+
+	if (structured)
+	{
+		solAssert(
+			structured->parameters() &&
+			structured->parameters()->parameters().size() == 1 &&
+			structured->parameters()->parameters().front() &&
+			*structured->parameters()->parameters().front()->annotation().type == *TypeProvider::stringMemory(),
+			""
+		);
+		solAssert(m_context.evmVersion().supportsReturndata(), "");
+
+		string const errorHash = FixedHash<4>(util::keccak256("Error(string)")).hex();
+
+		// Try to decode the error message.
+		// If this fails, leaves 0 on the stack, otherwise the pointer to the data string.
+		string const dataVariable = m_context.newYulVariable();
+		string const tryDecodeErrorMessage =
+			util::Whiskers(R"(
+				// try decoding the error message into <data>
+				let <data> := mload(0x40)
+				mstore(<data>, 0)
+				for {} 1 {} {
+					if lt(returndatasize(), 0x44) { <data> := 0 break }
```

Comment:
Shouldn't it revert on error?

Reply:
This part is a copy'n'paste from ContractCompiler.cpp's relevant part. hmm.